pulsar-transaction/coordinator/src/main/java/org/apache/pulsar/transaction/coordinator/impl/TxnLogBufferedWriter.java

```diff
@@ -0,0 +1,453 @@
+import org.apache.bookkeeper.util.SafeRunnable;
+import org.apache.commons.lang3.tuple.Triple;
+import org.apache.pulsar.common.allocator.PulsarByteBufAllocator;
+import org.apache.pulsar.common.util.collections.BitSetRecyclable;
+
+/***
+ * See PIP-160: https://github.com/apache/pulsar/issues/15516.
+ * Buffer requests and flush to Managed Ledger. Transaction Log Store And Pending Ack Store will no longer write to
+ * Managed Ledger directly, Change to using this class to write Ledger data.
+ * Caches “write requests” for a certain number or a certain size of request data and then writes them to the
+ * Managed Ledger in one go. After Managed Ledger has written complete, responds to each request-caller. In this
+ * process, Managed Ledger doesn't care how many records(or what to be written) in the Entry, it just treats them as
+ * a single block of data.
+ * The first write-request by transaction components will take a long time to receive a response, because we must wait
+ * for subsequent requests to accumulate enough data to actually start writing to the Managed Ledger. To control the
+ * maximum latency, we will mark the first request time for each batch, and additional timing triggers writes.
+ * You can enable or disabled the batch feature, will use Managed Ledger directly and without batching when disabled.
+ */
+@Slf4j
+public class TxnLogBufferedWriter<T> implements AsyncCallbacks.AddEntryCallback, Closeable {
+
+    public static final short BATCHED_ENTRY_DATA_PREFIX_MAGIC_NUMBER = 0x0e01;
+
+    public static final short BATCHED_ENTRY_DATA_PREFIX_VERSION = 1;
+
+    /**
+     * Enable or disabled the batch feature, will use Managed Ledger directly and without batching when disabled.
+     */
+    private final boolean batchEnabled;
+
+    private final ManagedLedger managedLedger;
+
+    private final OrderedExecutor orderedExecutor;
+
+    private final ScheduledExecutorService scheduledExecutorService;
+
+    /** The serializer for the object which called by {@link #asyncAddData}. **/
+    private final DataSerializer<T> dataSerializer;
+
+    private ScheduledFuture<?> scheduledFuture;
+
+    /**
+     * Caches “write requests” for a certain for a certain number, if reach this threshold, will trig Bookie writes.
+     */
+    private final int batchedWriteMaxRecords;
+
+    /**
+     * Caches “write requests” for a certain size of request data, if reach this threshold, will trig Bookie writes.
+     */
+    private final int batchedWriteMaxSize;
+
+    /** Maximum delay for writing to bookie for the earliest request in the batch. **/
+    private final int writeMaxDelayInMillis;
+
+    /** Data cached in the current batch. Will reset to null after each batched writes. **/
+    private ArrayList<T> dataArray;
+
+    /**
+     * Parameters of {@link #asyncAddData} cached in the current batch. Will reset to null after each batched writes.
+     */
+    private ArrayList<AsyncAddArgs> asyncAddArgsList;
```

Comment:
If the list is reset to null for each batched write, why not just create a new list for each batched write. So that we don't need to maintain a list here?

Reply:
And we should avoid creating a list for each batch? If there are 10k topics with 1ms max write delay, which means we will have 1000 * 10k ArrayList per second.

Reply:
> If the list is reset to null for each batched write, why not just create a new list for each batched write. 

Already fixed.

>  So that we don't need to maintain a list here? 

Transaction Log Buffered Write must hold either List or FlushContext because append requests are required to a container. 

> And we should avoid creating a list for each batch? If there are 10k topics with 1ms max write delay, which means we will have 1000 * 10k ArrayList per second.

I've changed the container for Transaction Log Buffered Write cache requests to `FlushContext`, and `FlushContext holds` reference of `ArrayList`, `ArrayList` will be reused by `FlushContext.recycle`.